### PR TITLE
Fix: Handle comments in block-spacing (fixes #4387)

### LINKED ISSUE #4387: Bug: Handle comments in block-spacing
`block-spacing` should handle comments the same way `space-in-parens` does, as discussed in #4302.

Here is an example of a valid case:
```js
{
  code: "if (a) {/* comment */ foo(); /* comment */}", 
  options: ["never"]
}
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/28159476-bug-handle-comments-in-block-spacing?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by alberto**

### CODE REVIEW

**Comment by nzakas** [comment]
Tests are failing.

File: lib/rules/block-spacing.js [hunk 1]
```diff
@@ -15,7 +15,8 @@ var util = require("../ast-utils");
 module.exports = function(context) {
     var always = (context.options[0] !== "never"),
         message = always ? "Requires a space" : "Unexpected space(s)",
-        sourceCode = context.getSourceCode();
+        sourceCode = context.getSourceCode(),
+        tokensAndComments = sourceCode.tokensAndComments;
```
Comment (mysticatea) [inline lib/rules/block-spacing.js]: Could we use TokenStore for `sourceCode.tokensAndComments`?
`context.getTokenAfter` and `context.getTokenBefore` are `O(1)` operation but `Array.prototype.indexOf` is `O(n)` operation.
And this `n` might be huge, and this operation is inside of other loop.

https://github.com/eslint/eslint/blob/master/lib/token-store.js

Comment (nzakas) [inline lib/rules/block-spacing.js]: That's a good point, though we would have to index each comment, which we don't currently do. And then we would need methods to get the comments (`getTokenOrCommentBefore`?)

Comment (mysticatea) [inline lib/rules/block-spacing.js]: Though I didn't confirm, I guessed `sourceCode.tokensAndComments = new TokenStore(ast.tokens.concat(ast.comments));` then `sourceCode.tokensAndComments.getTokenAfter(t)` might work.

**Comment by alberto** [comment]
@nzakas it's not that the tests are failing, but that our codebase is not compliant with this rule anymore [here](https://github.com/eslint/eslint/blob/master/lib/rules/vars-on-top.js#L103-L104). Should I fix that as part of the commit?

This happened yesterday to me and I fixed it. Then I got a different runtime error. I did an `npm install` just in case and the errors went away. `npm run check-commit && npm run docs` runs fine on my machine now.

File: lib/rules/block-spacing.js (hunk 1 above)
Comment (nzakas) [inline lib/rules/block-spacing.js]: We can't change what `tokensAndComments` is at this point. It needs to stay an array.

**Comment by nzakas** [comment]
@alberto I meant Travis was failing, not necessarily tests. It looks like linting is failing because of the changes you made to `block-spacing` in this commit. When you update a rule, you also need to make the related changes in the codebase so the build doesn't fail.

File: lib/rules/block-spacing.js (hunk 1 above)
Comment (alberto) [inline lib/rules/block-spacing.js]: Is it part of the public API now?

Is there anything else we can do? I could do what @mysticatea proposes inside the rule. Or we could add a new property and deprecate this one.

Comment (nzakas) [inline lib/rules/block-spacing.js]: Yes, SourceCode is part of the public API.  I don't see any reason to deprecate this, we just need a better way to get at individual tokens/comments than indexOf.

Comment (alberto) [inline lib/rules/block-spacing.js]: Is creating a TokenStore from the tokensAndComments inside the rule and use that ok for you?

File: lib/rules/block-spacing.js [hunk 2]
```diff
@@ -15,7 +16,9 @@ var util = require("../ast-utils");
 module.exports = function(context) {
     var always = (context.options[0] !== "never"),
         message = always ? "Requires a space" : "Unexpected space(s)",
-        sourceCode = context.getSourceCode();
+        sourceCode = context.getSourceCode(),
+        tokensAndComments = sourceCode.tokensAndComments,
+        tokenStore = new TokenStore(tokensAndComments);
```
Comment (nzakas) [inline lib/rules/block-spacing.js]: This is probably something we should be doing in `SourceCode` as part of the initialization. We really need `getTokenOrCommentBefore`, etc.

Comment (alberto) [inline lib/rules/block-spacing.js]: There are 13 methods in `TokenStore`. Do you want me to create the equivalent ones for all of them in `SourceCode`?

Comment (nzakas) [inline lib/rules/block-spacing.js]: No, I think we just need `getTokenOrCommentBefore` and `getTokenOrCommentAfter`.

Comment (alberto) [inline lib/rules/block-spacing.js]: Let me know if there is something else to do.

**Pull request merged**